scylla/src/transport/caching_session.rs

```diff
@@ -477,4 +498,61 @@ mod tests {
             assert!(session.prepare_batch(&bad_batch).await.is_err());
         }
     }
+
+    // The CachingSession::execute and friends should have the same StatementConfig
+    // and the page size as the Query provided as a parameter. It must not cache
+    // those parameters internally.
+    // Reproduces #597
+    #[tokio::test]
+    async fn test_parameters_caching() {
+        let session: CachingSession = CachingSession::from(new_for_test().await, 100);
+
+        session
+            .execute(
+                "CREATE TABLE tbl (pk int, ck int, PRIMARY KEY (pk, ck))",
+                (),
+            )
+            .await
+            .unwrap();
+
+        // Insert 10 rows
+        for i in 0..10 {
+            session
+                .execute("INSERT INTO tbl (pk, ck) VALUES (0, ?)", (i as i32,))
+                .await
+                .unwrap();
+        }
+
+        let q = Query::new("SELECT ck FROM tbl WHERE pk = ?");
+
+        // Request a page of size 3 first
+        let mut q3 = q.clone();
+        q3.set_page_size(3);
+
+        let rows3 = session
+            .execute_paged(q3.clone(), (0,), None)
+            .await
+            .unwrap()
+            .rows_typed::<(i32,)>()
+            .unwrap()
+            .collect::<Result<Vec<_>, _>>()
+            .unwrap();
+
+        assert_eq!(rows3, vec![(0,), (1,), (2,)]);
+
+        // Now request 7 rows
+        let mut q7 = q.clone();
+        q7.set_page_size(7);
+
+        let rows7 = session
+            .execute_paged(q7.clone(), (0,), None)
+            .await
+            .unwrap()
+            .rows_typed::<(i32,)>()
+            .unwrap()
+            .collect::<Result<Vec<_>, _>>()
+            .unwrap();
+
+        assert_eq!(rows7, vec![(0,), (1,), (2,), (3,), (4,), (5,), (6,)]);
+    }
```

Comment:
According to the protocol we don't, however we have some tests in Scylla which rely on that assumption so I think we should be safe for now.

I had a revelation just now - instead of checking `page_size`, we could check whether the `tracing` or `timestamp` parameter. I'll update the test to use one of them instead.